vowpalwabbit/debug_log.h

```diff
@@ -2,9 +2,50 @@
 // individual contributors. All rights reserved. Released under a BSD (revised)
 // license as described in the file LICENSE.
 
+#include "future_compat.h"
 #pragma once
 
-#define VW_DEBUG_ENABLE(x) static bool VW_DEBUG_LOG = x;
+struct vw_dbg
+{
+  static constexpr bool default_log         = false;
+  static constexpr bool learner             = false;
+  static constexpr bool search              = false;
+  static constexpr bool gd                  = false;
+  static constexpr bool gd_predict          = false;
+  static constexpr bool binary              = false;
+  static constexpr bool cb_adf              = false;
+  static constexpr bool csoaa               = false;
+  static constexpr bool cs_active           = false;
+  static constexpr bool cats_tree           = false;
+  static constexpr bool cats_pdf            = false;
+  static constexpr bool cats                = false;
+  static constexpr bool cb_explore_pdf      = false;
+  static constexpr bool cb_explore          = false;
+  static constexpr bool cbify               = false;
+  static constexpr bool get_pmf             = false;
+  static constexpr bool sample_pdf          = false;
+  static constexpr bool cb_explore_get_pmf  = false;
+  static constexpr bool cb_sample_pdf       = false;
+  static constexpr bool ccb                 = false;
+  static constexpr bool scorer              = false;
+  static constexpr bool cb_sample           = false;
 
-#define VW_DBG(e) \
-  if (VW_DEBUG_LOG) std::cout << depth_indent_string(e)
+  static constexpr bool track_stack =
+      default_log | learner     | search              | gd            | gd_predict |
+      binary      | cb_adf      | csoaa               | cs_active     | cats_tree  |
+      cats_pdf    | cats        | cb_explore_pdf      | cb_explore    | cbify      |
+      get_pmf     | sample_pdf  | cb_explore_get_pmf  | cb_sample_pdf | ccb        |
+      scorer      | cb_sample;
+};
+
+#define VW_DEBUG_LOG vw_dbg::default_log
+
+#define VW_LOG_SINK std::cout
+
+#define VW_DBG(e)                       \
+  if VW_STD17_CONSTEXPR (VW_DEBUG_LOG)  \
```

Comment:
if hasn't `STD17` what happens?

Reply:
The keyword constexpr will not get used since it won't be compatible with earlier compilers.